docs/topics/multiplatform/multiplatform-hierarchy.md

```diff
@@ -1,111 +1,322 @@
 [//]: # (title: Hierarchical project structure)
 
-With Kotlin 1.6.20, every new multiplatform project comes with a hierarchical project structure. This means that source
-sets form a hierarchy for sharing the common code among several targets. It opens up a variety of opportunities,
-including using platform-dependent libraries in common source sets and the ability to share code when creating multiplatform
-libraries.
-
-To get a default hierarchical project structure in your
-projects, [update to the latest release](releases.md#update-to-a-new-release). If you need to keep using an earlier version
-than 1.6.20, you can still enable this feature manually. For this, add the following to your `gradle.properties`:
-
-```none
-kotlin.mpp.enableGranularSourceSetsMetadata=true
-kotlin.native.enableDependencyPropagation=false
-```
+Multiplatform projects have support for hierarchical structure.
+This means you can form a hierarchy of source sets for sharing the common code among some, but not all,
+[supported targets](multiplatform-dsl-reference.md#targets).
```

Comment:
After this sentence, I immediately have a question: why would I want to share my code across some, but not all targets? 

The next few sentences kinda discuss it, but don't put it clearly enough. Let's try to clarify that.

Messaging:
Why do you need intermediate source-sets?

* If you're a library author, you might provide a specialized API for some, but not all targets (e.g. specialized API for K/N targets, but not for K/JVM targets). For that, you'll need to introduce an intermediate source set

* If you're a consumer of a library, you might want to use that specialized API, for that you'll have to introduce a similar intermediate source set

* Additionally, some libraries are not even available on some platforms. Specifically, Native Platform libraries are obviously available only in source sets that compile for K/Native only

* Kotlin Toolchain will ensure that each source set has access only to the API that is available on all targets to which this source set compiles. You're safe from using Windows-only API and then compiling it to MacOS, getting linkage errors or even undefined behaviour at runtime!

---

Note that the case "Sharing code in similar targets" mentioned in the original text is omitted. This is because IMO it's quite vague. Like I can theoretically come up with some case like "I use this utility function only in Native source-sets, thus I'll place it in allNative", but frankly, no one does that AFAIK. If you can write code in common, you write it in common and don't mess with hierarchies. 

That's a somewhat wild swing from the initial messaging of this article, so I'd like to hear your opinions on what do you think about adding those messages here. cc @KaterinaPetrova